library/std/src/thread/mod.rs

```diff
@@ -1001,10 +1001,13 @@ impl ThreadId {
         unsafe {
             let _guard = GUARD.lock();
 
-            // If we somehow use up all our bits, panic so that we're not
+            // If we somehow use up all our bits, abort so that we're not
             // covering up subtle bugs of IDs being reused.
             if COUNTER == u64::MAX {
-                panic!("failed to generate unique thread ID: bitspace exhausted");
+                drop(_guard); // in case this ends up calling `ThreadId::new()`, avoid reentrant lock acquire.
+                // Avoid panic machinery.
```

Comment:
It would be useful if there was a *why* in the code.

Reply:
If I understood @bjorn3 correctly, "because it might call this code reentrantly".

But if we do the `drop(_guard)` up above, is that even still a problem?

Reply:
The code might be called reentrantly if a panic is used using a custom panic handler. With an abort there is at least no stable way to reentrantly enter this function. Unstably maybe the stderr override used by eg libtest may be used to reentrantly enter this function.

Reply:
Right, but with the `drop`, reentrancy is no longer a problem, is it?

Reply:
Correct. The reentrant call can only happen after the mutex is unlocked.

Reply:
So then we can actually just make this raise a panic again?

Reply:
I think so.

Reply:
Okay, I switched the PR to that.